Basic login support (online-mode=false)

toqueteos opened the pull request:
Ready for merge :+1:
- Uses: Rust v1.4

fenhl commented on server/main.rs:
```diff
@@ -1,22 +1,411 @@
+    }
+
+    pub fn time_of_day(&self) -> i64 {
+        self.world_age() % 24000
```
That's not how time of day works.

toqueteos replied:
The same could be said about chunk columns, this might not be the way, but it works. There's no notion of world (there's just an enum for dimension) in the codebase, this way we can see what's needed and where.
EDIT: Actually, this is correct because there's no world persistence. But yes, it's wrong.

fenhl commented on server/main.rs:
```diff
@@ -1,22 +1,411 @@
+        let end = time::now().to_timespec();
+        let elapsed = end.sub(self.start).num_seconds();
+        elapsed * 20
+    }
```
That's not how world age works.

fenhl commented on server/main.rs:
```diff
@@ -1,22 +1,411 @@
+    }
+
+    pub fn handle(&self, mut stream: TcpStream) -> io::Result<()> {
+        let state = self.handle_handshake(&mut stream).unwrap();
```
Do we want a wrapper that disconnects clients with the panic message? Is that possible?

fenhl commented:
I'm still opposed to splitting up the `handle` function into separate parts for handshake and login. It does not reflect the linear nature of this part of the protocol.

fenhl commented:
Could we introduce a function or macro for writing a packet, flushing, and logging? The code is quite hard to read right now, and includes lots of repetition.

toqueteos commented:
Flushing _might be_ unnecessary most of the time (once at the end of handle should be enough!), it's there just to have some sort of _"breakpoints"_, so it's easy to see if which packet caused an error, not the best way of debugging of course.

**EDIT:** We seriously need to add [logging](https://crates.io/crates/log) ASAP.

toqueteos commented on src/proto/properties.rs:
```diff
@@ -0,0 +1,202 @@
+                "use-native-transport" => p.use_native_transport = value.parse().unwrap(),
+                "view-distance" => p.view_distance = value.parse().unwrap(),
+                "white-list" => p.white_list = value.parse().unwrap(),
+                prop => println!("Unknown property {}", prop),
```
Err instead of println!

toqueteos replied:
Actually... it may be good that it's a `println!`... We don't need to error out if a key is unknown, just log it to console. Right?

fenhl replied:
Yes, that's what vanilla does and I think it's a good idea to do the same in this case.

fenhl commented:
To facilitate multiworld support, client connections should be handled by a world instead of the server. The listener should pass new connections to a `handle` method on the default world instead of the server.

atheriel commented:
Why wouldn't you just keep track of the world the player was in?

fenhl commented on src/packet.rs:
```diff
@@ -467,7 +467,8 @@ pub mod login {
         0x01 => EncryptionRequest { server_id: String, pubkey: Arr<Var<i32>, u8>, verify_token: Arr<Var<i32>, u8> }
-        // 0x02 => LoginSuccess { uuid: Uuid, username: String; impl Protocol for LoginSuccess { ... } } // NOTE: uuid field is encoded as a string!
+        // NOTE: uuid field is encoded as a string!
+        0x02 => LoginSuccess { uuid: String, username: String }
```
Can't we use a `UuidString` type here which cleans up to a `Uuid`?

toqueteos replied:
Ofc, no problem! Curious: Is it used somewhere else?

fenhl replied:
Nope.

toqueteos replied:
Current draft: https://gist.github.com/toqueteos/4549b0b69eef4c1575c6

As atheriel pointed out on IRC, we could impl `FromError<uuid::ParseError>` for `io::Error`.

fenhl replied:
And as I pointed out, this doesn't need to be a custom packet if we can [add a UuidString type](https://www.irccloud.com/pastebin/tpCJMXqh/uuidstring.rs) instead.

fenhl commented on src/proto/ping.rs:
```diff
@@ -0,0 +1,102 @@
+        favicon: format!("data:image/png;base64,{:}", favicon),
+    };
+    let sr = packet::status::clientbound::StatusResponse{
+        response: json::encode(&resp).unwrap()
```
The JSON encoding should be handled by `StatusResponse::proto_encode`.

toqueteos replied:
Custom impl?

fenhl replied:
No, just impl `Protocol` for the `Response` struct and use that as the `StatusResponse` packet's field.

fenhl commented on src/proto/ping.rs:
```diff
@@ -0,0 +1,102 @@
+    // other values are static.
+    let resp = Response{
+        version: Version{
+            name: "1.8.1".to_string(),
```
1.8.3

fenhl commented on src/proto/ping.rs:
```diff
@@ -0,0 +1,102 @@
+    };
+    try!(sr.write(&mut stream));
+
+    // FIXME(toqueteos): Ping is optional, how we deal with that?
```
Loop over received packets, handle each status request and ping, regardless of order.

fenhl commented on src/proto/ping.rs:
```diff
@@ -0,0 +1,102 @@
+        wrong_packet => panic!("Expecting status::serverbound::Ping packet, got {:?}", wrong_packet)
+    };
+    let end = time::now().to_timespec();
+    let elapsed = end.sub(start).num_milliseconds();
```
[According to the wiki](http://wiki.vg/Protocol#Ping), that's not how ping works. The server should send back the exact same timestamp it received.

toqueteos replied:
I'm guessing this is for line 98 right? Yeah you are right, it should be `sp.time` not `elapsed`.

``` rust
// Original
let sp = packet::status::clientbound::Pong{time: elapsed};
// New 
let sp = packet::status::clientbound::Pong{time: sp.time};
```

Vanilla sends a pretty weird timestamp that's why elapsed is here, but it's not needed at all.

mindcat commented on server/main.rs:
```diff
@@ -1,22 +1,385 @@
+        }
+    }
+
+    pub fn handle_login(&self, mut stream: &mut TcpStream) -> io::Result<()> {
```
The process should be separate, controllable?

mindcat commented on server/main.rs:
```diff
@@ -1,22 +1,385 @@
+            max_players: 20,
+            level_type: "default".to_string(),
+            reduced_debug_info: false
+        }.write(stream));
```
Default value?
std::default::Default?
JoinGame::new?

fenhl replied:
Not required, this will sooner or later be replaced by actual values.

mindcat replied:
Okay :)

toqueteos committed:
Change server-port type i32 -> u16.

toqueteos committed:
Working vanilla server impl.

toqueteos commented:
Updated to latest master.
- Handles missing `server.properties` files.
- I've moved almost everything from `main.rs` into `src/vanilla/server.rs`. There are leftover comments, I tried to move what's left of main.rs into a `run` method inside `Server` but we need an `Rc`/`Arc` for `self` then.
- It now uses `log` crate. Most `println!` messages are now `debug!` (which by default shouldn't be shown) all others are `info!`.
- As extra I implemented packet `ChunkDataBulk` (0x26) and modified `ChunkColumn` decoding to work nicely with both 0x21 and 0x26.
- `server-port` key from `server.properties` is now read as an `u16` not an `i32`.

EDIT: And Travis doesn't like `cargo doc` for log crate so it failed, YAY!

atheriel commented:
I would be ok with merging this if the docs for the log crate was the only bottleneck.

fenhl commented on src/packet.rs:
```diff
@@ -352,7 +358,59 @@ pub mod play {
+            impl Protocol for ChunkDataBulk {
+                type Clean = Self;
+                fn proto_len(this: &Self) -> usize {
+                    let mut len = 1; // sky_light_sent(bool) len is constant
```
Mutability is absolutely not needed here.

toqueteos replied:
How do you modify len at lines 367, 369 and 372 then?

fenhl replied:
``` rust
fn proto_len(this: &Self) -> usize {
    let columns = this.chunk_meta.len() as i32;
    1 // sky_light_sent(bool) len is constant
    + <Var<i32> as Protocol>::proto_len(&columns)
    + this.chunk_meta.iter().map(|cm| <ChunkMeta as Protocol>::proto_len(cm)).sum()
    + this.chunk_data.iter().map(|cd| cd.len()).sum()
}
```

Requires the [`AdditiveIterator`](http://doc.rust-lang.org/std/iter/trait.AdditiveIterator.html) trait.

toqueteos replied:
Applied in #53.

fenhl commented on src/packet.rs:
```diff
@@ -352,7 +358,59 @@ pub mod play {
+                    let mut data = Vec::with_capacity(1 << 12);
+                    try!(src.read_to_end(&mut data));
+                    let mut chunk_data = Vec::with_capacity(columns as usize);
+                    for (idx, cd) in chunk_data.iter_mut().enumerate() {
```
Can we zip `chunk_data` and `chunk_meta` instead of enumerating?

toqueteos replied:
Applied in #53.

fenhl commented on src/packet.rs:
```diff
@@ -352,7 +358,59 @@ pub mod play {
+                    let mut chunk_data = Vec::with_capacity(columns as usize);
+                    for (idx, cd) in chunk_data.iter_mut().enumerate() {
+                        // chunk_data, mask, continuous, sky_light
+                        *cd = try!(ChunkColumn::decode(data.clone(), chunk_meta[idx].mask, true, true));
```
Would it be possible for `ChunkColumn::decode` to take a reference to `data`, to avoid the cloning?

toqueteos replied:
Fixed in #53.

fenhl commented on src/vanilla/server.rs:
```diff
@@ -0,0 +1,337 @@
+            }
+        };
+        Ok(())
+    }
```
`handle_handshake`, `handle_login`, and `handle_post_login` still need to be moved into this function. Also it would be better for the `NextState::Status` branch to loop over incoming packets and handle pings and status requests in any order.

toqueteos commented:
Rebased onto master, much shorter now. Same Travis error.

fenhl commented on src/lib.rs:
```diff
@@ -1,11 +1,15 @@
 extern crate flate2;
+#[macro_use]
+extern crate log;
 extern crate nbt;
```
missing `extern crate rand;`

toqueteos replied:
Fixed!

toqueteos commented:
Good to go, no more weird log crate errors.

toqueteos commented:
As per @fenhl 's request, I moved all `handle_*` fn logic into `handle`. I still don't get the benefits of this change in the long-term but there it is.

toqueteos commented:
Most issues solved. There's tons of TODO and FIXME comments which will get resolved in follow-up PRs :tada:

fenhl commented on src/world.rs:
```diff
@@ -0,0 +1,246 @@
+use rand;
+use time;
+
+const PACKET_NAMES: [&'static str; 26] = [
```
Can we get rid of this by making packets `Debug`?

**Edit:** it's probably fine if we do that in a follow-up PR.

toqueteos replied:
We probably need to do it in a follow-up PR because there were some weird types not having Debug so we have to manually implement it.

fenhl commented:
Looks good to merge. :+1:

bvssvni commented:
Please make a blog post, or post something on r/rust_gamedev.

toqueteos commented:
Build broken because of `Error` trait stabilization (`rustc_serialize` and `byteorder`). No big worries, already adressed, just waiting for new nightlies.

toqueteos commented:
I'll revisit this very soon. Rebase + upgrade to latest rust (v1.4)

fenhl commented on server/main.rs:
```diff
@@ -1,20 +1,59 @@
+}
+
+fn main () {
+    init().unwrap();
```
This should have a more descriptive name. `init_logger` or something. And all of the `unwrap` calls should be `expect` instead.

toqueteos replied:
https://github.com/PistonDevelopers/hematite_server/commit/0166367333c3eb1ddc32ee8fa50cec4d5e1a4f7f

toqueteos committed:
Use expect instead of unwrap

toqueteos commented:
@fenhl @bvssvni ping?

bvssvni commented:
Looks good!

toqueteos commented:
Done, we are one step closer to hematite being all Rust!
